reactor-core/src/main/java/reactor/core/scheduler/BoundedElasticScheduler.java

```diff
@@ -319,492 +244,554 @@ public String toString() {
+		  .append(",ttl=");
+		if (ttlMillis < 1000) {
+			ts.append(ttlMillis).append("ms)");
+		}
+		else {
+			ts.append(ttlMillis / 1000).append("s)");
+		}
 		return ts.toString();
 	}
 
+	/**
+	 * @return a best effort total count of the spinned up executors
+	 */
+	int estimateSize() {
+		return BOUNDED_SERVICES.get(this).get();
+	}
+
+	/**
+	 * @return a best effort total count of the busy executors
+	 */
+	int estimateBusy() {
+		return BOUNDED_SERVICES.get(this).busyQueue.size();
+	}
+
+	/**
+	 * @return a best effort total count of the idle executors
+	 */
+	int estimateIdle() {
+		return BOUNDED_SERVICES.get(this).idleQueue.size();
+	}
+
+	/**
+	 * Best effort snapshot of the remaining queue capacity for pending tasks across all the backing executors.
+	 *
+	 * @return the total task capacity, or {@literal -1} if any backing executor's task queue size cannot be instrumented
+	 */
+	int estimateRemainingTaskCapacity() {
+		Queue<BoundedState> busyQueue = BOUNDED_SERVICES.get(this).busyQueue;
+		int totalTaskCapacity = maxTaskQueuedPerThread * maxThreads;
+		for (BoundedState state : busyQueue) {
+			int stateQueueSize = state.estimateQueueSize();
+			if (stateQueueSize >= 0) {
+				totalTaskCapacity -= stateQueueSize;
+			}
+			else {
+				return -1;
+			}
+		}
+		return totalTaskCapacity;
+	}
+
 	@Override
 	public Object scanUnsafe(Attr key) {
 		if (key == Attr.TERMINATED || key == Attr.CANCELLED) return isDisposed();
-		if (key == Attr.CAPACITY) return threadCap;
-		if (key == Attr.BUFFERED) return idleServicesWithExpiry.size(); //BUFFERED: number of workers alive and backed by thread
+		if (key == Attr.BUFFERED) return estimateSize();
+		if (key == Attr.CAPACITY) return maxThreads;
 		if (key == Attr.NAME) return this.toString();
 
 		return null;
 	}
 
 	@Override
-		//TODO re-evaluate the inners? should these include deferredWorkers? allServices?
 	public Stream<? extends Scannable> inners() {
-		return idleServicesWithExpiry.stream()
-		                             .map(cached -> cached.cached);
+		BoundedServices services = BOUNDED_SERVICES.get(this);
+		return Stream.concat(services.busyQueue.stream(), services.idleQueue.stream())
+		             .filter(obj -> obj != null && obj != CREATING);
 	}
 
-	void eviction(LongSupplier nowSupplier) {
-		long now = nowSupplier.getAsLong();
-		List<CachedServiceExpiry> list = new ArrayList<>(idleServicesWithExpiry);
-		for (CachedServiceExpiry e : list) {
-			if (e.expireMillis < now) {
-				if (idleServicesWithExpiry.remove(e)) {
-					e.cached.exec.shutdownNow();
-					allServices.remove(e.cached);
-					REMAINING_THREADS.incrementAndGet(this);
-				}
-			}
-		}
+	@Override
+	public Worker createWorker() {
+		BoundedState picked = BOUNDED_SERVICES.get(this)
+		                                      .pick();
+		ExecutorServiceWorker worker = new ExecutorServiceWorker(picked.executor);
+		worker.tasks.add(picked); //this ensures the BoundedState will be released when worker is disposed
```

Comment:
unrelated nit: `ExecutorServiceWorker#tasks` is very misleading (especially when populated from places like here), `ExecutorServiceWorker#disposables` or something would make it much easier to understand what it actually does (since it does not execute `tasks`, only disposes them)

Reply:
used to bear a reference to the `Disposable` handle that allows cancelling all the tasks, but can contain more than just task-related handlers. will rename 👍